tests/kernels/test_causal_conv1d.py

```diff
@@ -190,16 +196,96 @@ def test_causal_conv1d_update(batch, dim, width, has_bias, silu_activation,
                                conv_state,
                                weight,
                                bias,
-                               activation=activation)
+                               activation=activation,
+                               cache_seqlens=cache_seqlens)
     out_ref = causal_conv1d_update_ref(x,
                                        conv_state_ref,
                                        weight,
                                        bias,
-                                       activation=activation)
+                                       activation=activation,
+                                       cache_seqlens=cache_seqlens)
 
     assert torch.equal(conv_state, conv_state_ref)
     assert torch.allclose(out, out_ref, rtol=rtol, atol=atol)
+
+
+@pytest.mark.parametrize("itype", [torch.bfloat16])
+@pytest.mark.parametrize("silu_activation", [True])
+@pytest.mark.parametrize("has_bias", [True])
+@pytest.mark.parametrize("width", [4])
+@pytest.mark.parametrize('seqlen',
+                         [8, 16, 32, 64, 128, 256, 512, 784, 1024, 2048, 4096])
+@pytest.mark.parametrize('dim', [64, 4096])
+def test_causal_conv1d_varlen(dim, seqlen, width, has_bias, silu_activation,
+                              itype):
+    device = "cuda"
+    rtol, atol = (3e-4, 1e-3) if itype == torch.float32 else (3e-3, 5e-3)
+    if itype == torch.bfloat16:
+        rtol, atol = 1e-2, 5e-2
+    # set seed
+    torch.random.manual_seed(seqlen + dim + width)
+    batch = 1
+    seqlens = []
+    nsplits = 3
+    eos_pos = torch.randperm(seqlen - 1)[:nsplits].sort().values
+    seqlens.append(
+        torch.diff(
+            torch.cat(
+                [torch.tensor([-1]), eos_pos,
+                 torch.tensor([seqlen - 1])])).tolist())
+    assert sum(seqlens[-1]) == seqlen
+    assert all(s > 0 for s in seqlens[-1])
+
+    cumsum = torch.cumsum(torch.tensor(seqlens[0]), dim=0).to(torch.int32)
+    x = torch.randn(batch, 4096 + dim + 64, seqlen, device=device,
+                    dtype=itype)[:, 4096:4096 + dim, :]
+    weight = torch.randn(dim, width, device=device, dtype=itype)
+    bias = torch.randn(dim, device=device, dtype=itype) if has_bias else None
+    x_ref = x.clone()
+    weight_ref = weight.clone()
+    bias_ref = bias.clone() if bias is not None else None
+    activation = None if not silu_activation else "silu"
+    final_states = torch.randn(nsplits + 1,
+                               dim,
+                               width - 1,
+                               device=x.device,
+                               dtype=x.dtype)
+    final_states_ref = final_states.clone()
+    has_initial_states = torch.ones_like(cumsum,
+                                         dtype=torch.int32,
+                                         device=x.device)
+    cache_indices = torch.arange(cumsum.shape[0],
+                                 dtype=torch.int32,
+                                 device=x.device)
```

Comment:
Is there a test where we randomize the cache indices? Could you add this if possible?

Reply:
Added it, Thanks!